File: bin/benchmark/json-diagnostic/owned-reference/lifecycle.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"runtime"
	"time"

	rt "gopurs/output/gopurs_runtime"
	p "gopurs/output/purescript"
)

type sample struct {
	TimeUS         float64 `json:"time_us"`
	AllocatedBytes uint64  `json:"allocated_bytes"`
}

func measure(action func()) sample {
	var before, after runtime.MemStats
	runtime.ReadMemStats(&before)
	start := time.Now()
	action()
	elapsed := time.Since(start).Nanoseconds()
	runtime.ReadMemStats(&after)
	return sample{float64(elapsed) / 1000, after.TotalAlloc - before.TotalAlloc}
}

func main() {
	var files []struct {
		Name, Contents string
		Benchmark      bool
	}
	bytes, err := os.ReadFile(os.Getenv("DIAG_CORPUS"))
	if err != nil {
		panic(err)
	}
	if err = json.Unmarshal(bytes, &files); err != nil {
		panic(err)
	}
	inputs, names := []string{}, []string{}
	for _, file := range files {
		if file.Benchmark {
			inputs = append(inputs, file.Contents)
			names = append(names, file.Name)
		}
	}
	// Clear input-loader scratch before constructing any decoder or result.
	runtime.GC()
	var combined func(string) rt.Value
	construction := measure(func() {
		// BUILD_DECODER
	})
	first, second := make([]rt.Value, len(inputs)), make([]rt.Value, len(inputs))
	pass := func(results []rt.Value) sample {
		return measure(func() {
			for i, input := range inputs {
				results[i] = combined(input)
			}
		})
	}
	firstSample, secondSample := pass(first), pass(second)
	// Charge deferred parser/decoder garbage before creating any fingerprint
	// scratch. Both complete batches remain live through this timed collection.
	retainedCollection := measure(runtime.GC)
	var retained, released runtime.MemStats
	runtime.ReadMemStats(&retained)
	fingerprint := p.Get_Test_JsonDecoding_fingerprint()
	hashes := func(results []rt.Value) []string {
		out := make([]string, len(results))
		for i, result := range results {
			var value any
			if err := json.Unmarshal([]byte(rt.Apply(fingerprint, result).StrVal()), &value); err != nil {
				panic(err)
			}
			bytes, err := json.Marshal(value)
			if err != nil {
				panic(err)
			}
			hash := sha256.Sum256(bytes)
			out[i] = hex.EncodeToString(hash[:])
		}
		return out
	}
	firstHashes, secondHashes := hashes(first), hashes(second)
	// Only validation scratch is now excluded. Decoding garbage was charged
	// above, and result reclamation is charged below after dropping both roots.
	runtime.GC()
	runtime.KeepAlive(first)
	runtime.KeepAlive(second)
	first, second = nil, nil
	drain := measure(runtime.GC)
	runtime.ReadMemStats(&released)
	if err := json.NewEncoder(os.Stdout).Encode(map[string]any{
		"names": names, "construction": construction,
		"first_combined": firstSample, "second_combined": secondSample,
		"retained_collection": retainedCollection, "release_collection": drain,
		"retained_heap_bytes": retained.HeapAlloc, "released_heap_bytes": released.HeapAlloc,
		"first_fingerprints": firstHashes, "second_fingerprints": secondHashes,
	}); err != nil {
		panic(err)
	}
	runtime.KeepAlive(inputs)
	runtime.KeepAlive(combined)
}

File: bin/benchmark/json-diagnostic/owned-reference/lifecycle.cc
// Fresh-process construction/two retained decodes/destruction audit.
#define main reference_benchmark_main
#ifdef REFERENCE_TAST
#include "../typed-ast.cc"
#else
#include "../decoding.cc"
#endif
#undef main

int main() {
  std::string text;
  if (!read_text(getenv("DIAG_CORPUS"), text)) return 2;
  simdjson::dom::parser corpus_parser;
  std::vector<Case> cases;
  for (element raw : array_of(corpus_parser.parse(text).value())) {
    object input = object_of(raw);
#ifndef REFERENCE_TAST
    if (!bool_of(require_field(input, "benchmark"))) continue;
#endif
    Case item;
    item.name = string_of(require_field(input, "name"));
    item.contents = string_of(require_field(input, "contents"));
    cases.push_back(std::move(item));
  }
#ifdef REFERENCE_TAST
  using Result = Module;
#else
  using Result = Payload;
#endif
  auto decode = [](const std::string &input) {
    simdjson::dom::parser parser;
    element root = parser.parse(input).value();
#ifdef REFERENCE_TAST
    return decode_module(root, "");
#else
    Payload value;
    decode_payload(root, value);
    return value;
#endif
  };
  // There is no decoder/getter state to preconstruct in this C++ API. Parser
  // construction/allocation/destruction is included in each full decode.
  std::vector<Result> first(cases.size()), second(cases.size());
  auto pass = [&](std::vector<Result> &outputs) {
    double begin = now_microseconds();
    for (size_t index = 0; index < cases.size(); index++) outputs[index] = decode(cases[index].contents);
    return now_microseconds() - begin;
  };
  double first_us = pass(first), second_us = pass(second);
  auto hashes = [](const std::vector<Result> &outputs) {
    std::string out = "[";
    for (size_t index = 0; index < outputs.size(); index++) {
      if (index) out += ',';
#ifdef REFERENCE_TAST
      out += json_quoted(sha256_hex(module_fingerprint(outputs[index])));
#else
      out += json_quoted(payload_fingerprint(outputs[index]));
#endif
    }
    return out + ']';
  };
  std::string first_hashes = hashes(first), second_hashes = hashes(second);
  double begin = now_microseconds();
  first.clear(); second.clear();
  double release_us = now_microseconds() - begin;
  printf("{\"construction\":{\"time_us\":0},\"first_combined\":{\"time_us\":%.9f},"
         "\"second_combined\":{\"time_us\":%.9f},\"retained_collection\":{\"time_us\":0},"
         "\"release_collection\":{\"time_us\":%.9f},"
         "\"first_fingerprints\":%s,\"second_fingerprints\":%s}\n",
         first_us, second_us, release_us, first_hashes.c_str(), second_hashes.c_str());
}
